### PR TITLE
Revert HostZoneUnbonding Status upon Channel Restoration

### PR DESCRIPTION
**Pull request by sampocs**
Closes: #XXX

## Context and purpose of the change
After a channel has been closed, it's possible for records to be stuck in an `IN_PROGRESS` state. This PR reverts the status during a channel restoration.

## Brief Changelog
* Reverts the host zone unbonding record status from: `EXIT_TRANSFER_IN_PROGRESS` to` EXIT_TRANSFER_QUEUE`, and `UNBONDING_IN_PROGRESS` to `UNBONDING_QUEUE` when the delegation ICA channel is restored. 
* Added unit tests for epoch unbonding records
* Modified `SetHostZoneUnbondings` to take a `ChainId` as a parameter instead of a `HostZone` struct

#### Network Startup
* Add `--block-history 1000` to the command in `Dockerfile.relayer`, like so:
```
RUN echo 'rly start $1 -p events --block-history 1000' > start.sh
```
* Disable the other host zones by setting the `HOST_CHAINS` array in `scripts/start_network.sh` to only `GAIA`
* Start dockenet
```
git submodule update --init --recursive
make start-docker build=sgr
```

#### Transaction Startup
* In a new terminal window, run the following to continuously liquid stake
```bash
bash scripts/test-util/1.sh 
while true; do bash scripts/test-util/2.sh && sleep 30; done
```
* Wait for the stake to land on the host, and then run the following in a new window to continuously trigger unbondings
```bash
while true; do bash scripts/test-util/3.sh && sleep 30; done
```

#### Close the Channel
* Wait for an unbonding to process, identified by when an epoch unbonding record status is set to CLAIMABLE.  You can view all the records in`scripts/logs/state.log`
* Then modify the go relayer in `scripts/state/relayer-gaia/config/config.yaml` to ignore the ICA channels
```bash
sed -i -E 's|rule: ""|rule: "allowlist"|g' scripts/state/relayer-gaia/config/config.yaml
sed -i -E 's|channel-list: \[\]|channel-list: \[channel-0\]|g' scripts/state/relayer-gaia/config/config.yaml
```

* Restart the go relayer to see the changes take effect
```
docker rm -f stride-relayer-gaia-1
docker-compose up relayer-gaia
```
* Wait a few minutes to let records pile up. Ensure there are multiple records in the following state before proceeding:
  * deposit records in state `DELEGATION_IN_PROGRESS` 
  * host zone unbonding records in state `UNBONDING_IN_PROGRESS` 
  * host zone unbonding records in state `EXIT_TRANSFER_IN_PROGRESS` 

#### Restore the Channel
* Revert the go relayer back to its original settings and restart it again 
```bash
sed -i -E 's|rule: "allowlist"|rule: ""|g' scripts/state/relayer-gaia/config/config.yaml
sed -i -E 's|channel-list: \[channel-0\]|channel-list: \[\]|g' scripts/state/relayer-gaia/config/config.yaml

docker rm -f stride-relayer-gaia-1
docker-compose up relayer-gaia
```
* Confirm the delegation channel and withdrawal channel are closed [here](http://localhost:1317/ibc/core/channel/v1/channels)
* View the deposit records and epoch unbonding records at `scripts/logs/state.log`
* Restore the withdrawal channel and confirm records HAVE NOT changed
```
build/strided --home scripts/state/stride1 tx stakeibc restore-interchain-account GAIA WITHDRAWAL --from val1 --gas 400000
```
* Restore the delegation channel and confirm that the deposit records and host zone unbonding records were reverted 
```
build/strided --home scripts/state/stride1 tx stakeibc restore-interchain-account GAIA DELEGATION --from val1 --gas 400000
```

## Author's Checklist

I have...

- [ ] Run and PASSED locally all GAIA integration tests
- [ ] If the change is contentful, I either:
    - [x] Added a new unit test OR 
    - [x] Added test cases to existing unit tests
- [ ] OR this change is a trivial rework / code cleanup without any test coverage

If skipped any of the tests above, explain.

## Reviewers Checklist

*All items are required. Please add a note if the item is not applicable and please add
your handle next to the items reviewed if you only reviewed selected items.*

I have...

- [ ] reviewed state machine logic
- [ ] reviewed API design and naming
- [ ] manually tested (if applicable)
- [ ] confirmed the author wrote unit tests for new logic
- [ ] reviewed documentation exists and is accurate

## Documentation and Release Note

  - [ ] Does this pull request introduce a new feature or user-facing behavior changes? 
  - [ ] Is a relevant changelog entry added to the `Unreleased` section in `CHANGELOG.md`?
  - [ ] This pull request updates existing proto field values (and require a backend and frontend migration)? 
  - [ ] Does this pull request change existing proto field names (and require a frontend migration)?
  How is the feature or change documented? 
      - [ ] not applicable
      - [ ] jira ticket `XXX` 
      - [ ] specification (`x/<module>/spec/`) 
      - [ ] README.md 
      - [ ] not documented

### CODE REVIEW

File: x/stakeibc/keeper/msg_server_restore_interchain_account.go [hunk 1]
```diff
@@ -44,15 +44,58 @@ func (k msgServer) RestoreInterchainAccount(goCtx context.Context, msg *types.Ms
...
+		recordsStuckInTransfer := []uint64{}
+		for _, epochUnbondingRecord := range epochUnbondingRecords {
 			// only revert records for the select host zone
-			if record.HostZoneId == hostZone.ChainId && record.Status == recordtypes.DepositRecord_DELEGATION_IN_PROGRESS {
-				record.Status = recordtypes.DepositRecord_DELEGATION_QUEUE
-				k.Logger(ctx).Error(fmt.Sprintf("Setting DepositRecord %d to status DepositRecord_DELEGATION_IN_PROGRESS", record.Id))
-				k.RecordsKeeper.SetDepositRecord(ctx, record)
+			hostZoneUnbonding, found := k.RecordsKeeper.GetHostZoneUnbondingByChainId(ctx, epochUnbondingRecord.EpochNumber, hostZone.ChainId)
+			if !found {
+				continue
```
Comment (asalzmann) [inline x/stakeibc/keeper/msg_server_restore_interchain_account.go:66]:
```suggestion
                                 k.Logger(ctx).Info(fmt.Sprintf("No HostZoneUnbonding found for chainId: %s, epoch: %d", hostZone.ChainId, epochUnbondingRecord.EpochNumber))
				continue
```

File: x/stakeibc/keeper/msg_server_restore_interchain_account.go [hunk 2]
```diff
@@ -44,15 +44,58 @@ func (k msgServer) RestoreInterchainAccount(goCtx context.Context, msg *types.Ms
...
+			if depositRecord.HostZoneId == hostZone.ChainId && depositRecord.Status == recordtypes.DepositRecord_DELEGATION_IN_PROGRESS {
+				depositRecord.Status = recordtypes.DepositRecord_DELEGATION_QUEUE
+				k.Logger(ctx).Info(fmt.Sprintf("Setting DepositRecord %d to status DepositRecord_DELEGATION_IN_PROGRESS", depositRecord.Id))
+				k.RecordsKeeper.SetDepositRecord(ctx, depositRecord)
+			}
+		}
+
+		// revert epoch unbonding records for the closed ICA channel
+		epochUnbondingRecords := k.RecordsKeeper.GetAllEpochUnbondingRecord(ctx)
+		recordsStuckInUnbonding := []uint64{}
```
Comment (asalzmann) [inline x/stakeibc/keeper/msg_server_restore_interchain_account.go:60]: nit: should we call this `epochNumberForPendingRecord`  or something similar?

**Review by asalzmann — APPROVED** [review]
Lgtm! Thanks for taking this ticket, very thorough unittests

File: x/stakeibc/keeper/msg_server_restore_interchain_account.go (hunk 2 above)
Comment (sampocs) [reply x/stakeibc/keeper/msg_server_restore_interchain_account.go:60]: yup, good call, updating

**Pull request merged**